File: 8-queue-and-stack/min-string-after-remove.py
"""
Leet code 2696: Minimum String After Removing Substrings
Techniques: stack, greedy
Complexity: O(n) where n is the length of the string
Space Complexity: O(n) for the stack
"""


class Solution(object):
    def minLength(self, s):
        stack = []
        for char in s:
            if stack and stack[-1] + char in ("AB", "CD"):
                stack.pop()
            else:
                stack.append(char)
        return len(stack)

File: 8-queue-and-stack/min-string-after-remove.js
/**
 * @param {string} s
 * @return {number}
 */
var minLength = function(s) {
    let stack = []
    for (let char of s){
        if (stack.length > 0 && ['AB','CD'].includes(stack.at(-1)+char)) {
            stack.pop()
        } else {
            stack.push(char)
        }
    }
    return stack.length
};